forge/test/mlir/llama/test_llama_inference.py

```diff
@@ -120,3 +121,346 @@ def test_llama_inference_cache_cpu():
+                    torch.zeros(
+                        v.shape[-4], v.shape[-3], max_new_tokens, v.shape[-1]
+                    ).to(k.dtype),
+                ],
+                dim=-2,
+            )
+
+    # # Compile the model
+    # compiled_model = forge.compile(llama_model, sample_inputs=model_inputs)
+
+    # Run decode stage on TT device and generate tokens by passing the last predicted token and the past key values.
+    # untill the a specified maximum number of new tokens is reached or an end-of-sequence token is encountered.
+    for max_new_tokens_idx in range(max_new_tokens):
+
+        # CPU Inference
+        model_outputs = llama_model(model_inputs[0], model_inputs[1])
+
+        # TT will return the logits and past key values as list of tensor, so flattening
+        # framework output past key values from List(List(Key1, Values1), ... , List(Key26, Values26)) to
+        # List(Key1, Values1, ... , Key26, Values26) for comparing the Framework and TT output in similar fashion.
+        framework_output = [model_outputs[0]]
+        for k, v in model_outputs[1]:
+            framework_output.append(k)
+            framework_output.append(v)
+
+        # # Run on TT device
+        # tt_output = compiled_model(*model_inputs)
+        # tt_output = [tt_out.to("cpu") for tt_out in tt_output]
+
+        # # Validate TT result with Framework
+        # assert all([compare_with_golden_pcc(golden=fw_out, calculated=tt_out, pcc=0.99) for fw_out, tt_out in zip(framework_output, tt_output)])
+
+        logits = framework_output[0]
+        past_key_values = framework_output[1:]
+
+        next_token_logits = logits[:, -1, :]
+        next_token = torch.argmax(next_token_logits, dim=-1)
+        print("next_token=", next_token)
+
+        if next_token == tokenizer.eos_token_id:
+            break
+
+        model_inputs = [next_token.unsqueeze(0)]
+
+        # Formate the past key values from List(Key1, Values1, ... , Key26, Values26) to
+        # List(List(Key1, Values1), ... , List(Key26, Values26))
+        model_inputs.append(
+            [
+                past_key_values[idx : idx + 2]
+                for idx in range(0, len(past_key_values), 2)
+            ]
+        )
+
+        if apply_padding_on_past_key_values:
+
+            for idx in range(len(model_inputs[1])):
+                key_state = torch.clone(model_inputs[1][idx][0][:, :, -1, :])
+                value_state = torch.clone(model_inputs[1][idx][1][:, :, -1, :])
+                model_inputs[1][idx][0][
+                    :, :, non_padding_seq_len + max_new_tokens_idx, :
+                ] = model_inputs[1][idx][0][:, :, -1, :]
+                model_inputs[1][idx][0] = model_inputs[1][idx][0][:, :, :-1, :]
+                model_inputs[1][idx][1][
+                    :, :, non_padding_seq_len + max_new_tokens_idx, :
+                ] = model_inputs[1][idx][1][:, :, -1, :]
+                model_inputs[1][idx][1] = model_inputs[1][idx][1][:, :, :-1, :]
```

Comment:
Here, I am taking the last key_states and values_states that are concatenated at the end of the past key values in LlamaAttention to the first non padding index in past key values.

Reply:
Nice